course/permissions.py

```diff
@@ -123,13 +145,18 @@ def is_object_visible(self, request, view, course_instance):
 class CourseModulePermission(MessageMixin, Permission):
     message = _('MODULE_PERMISSION_MSG_NOT_VISIBLE')
 
-    def has_permission(self, request, view):
+    def has_permission(self, request: HttpRequest, view: 'CourseModuleMixin') -> bool:
```

Comment:
Why is CourseModuleMixin a string?

Reply:
This seems to happen in quite a few places

Reply:
The same reason CourseModuleMixin is imported like this:
https://github.com/apluslms/a-plus/blob/d95abeaabb607eccf9c8c5b920e1b9c4def23aa7/course/permissions.py#L22-L23
In some cases, importing classes for type annotation purposes can cause circular imports (file A imports some classes from file B, which imports some classes from file A). This causes an exception at runtime. The workaround for this is to import the classes when `TYPE_CHECKING` is true, and refer to it as a string in type annotations. The type checker works the same way regardless of if the type annotation is a string.

I'm not sure if this would've caused a circular import *in this particular case*, but in many cases they did. For consistency, I decided to use this style for all classes that were imported for annotation purposes only.

Reply:
Oh, I completely forgot that you need to have it as a string due to it not existing during runtime.